src/sqlfluff/rules/structure/ST03.py

```diff
@@ -23,7 +23,8 @@ def _is_data_modifying_cte(cte_query: Query) -> bool:
         except StopIteration:
             return False
 
-    return False
+    # raise NotImplementedError("CTE definition is required")
```

Comment:
When I add this line, I get a coverage violation. When I remove it, I get a MyPy violation. Should I do something like `assert_never`?